### PR TITLE
[frost] new_keygen safety

### PR DESCRIPTION
**Pull request by LLFourn**
Make it safer by encouraging passing in your secret polynomial directly.

### CODE REVIEW

File: schnorr_fun/src/frost.rs [hunk 1]
```diff
@@ -661,14 +663,34 @@ impl<H: Digest<OutputSize = U32> + Clone, NG> Frost<H, NG> {
...
         &self,
-        point_polys: BTreeMap<PartyIndex, Vec<Point>>,
-    ) -> Result<KeyGen, NewKeyGenError> {
+        mut point_polys: BTreeMap<PartyIndex, Vec<Point>>,
+        local_secret_polys: &BTreeMap<PartyIndex, S>,
+    ) -> Result<KeyGen, NewKeyGenError>
+    where
+        S: AsRef<[Scalar]>,
+    {
+        for (party_id, scalar_poly) in local_secret_polys {
```
Comment (nickfarrow) [inline schnorr_fun/src/frost.rs:687]: Is this futureproofing for having a single party control multiple shares of the FROST key?

File: schnorr_fun/src/frost.rs [hunk 2]
```diff
@@ -586,12 +592,8 @@ impl<H: Digest<OutputSize = U32> + Clone, NG: NonceGen> Frost<H, NG> {
                 )
             })
             .collect::<BTreeMap<_, _>>();
-        let point_polys = scalar_polys
-            .iter()
-            .map(|(party_index, sp)| (*party_index, to_point_poly(sp)))
-            .collect::<BTreeMap<_, _>>();
 
-        let keygen = self.new_keygen(point_polys).unwrap();
+        let keygen = self.new_keygen(Default::default(), &scalar_polys).unwrap();
```
Comment (nickfarrow) [inline schnorr_fun/src/frost.rs:596]: 👍

**Review by nickfarrow — APPROVED** [review]
Looks good!

This risk was introduced by us being lazy not passing our public poly into finish keygen code. Just receiving a `FinishKeygen` message from the coordinator which contains all the public polys including our own.

Others will likely do the same, so definitely worth highlighting.

It does make `new_keygen` slightly clunkier, skipping the check with `BTreeMap::new() or Default::default()`, but i think this is probably nicer than an `Option<BTreeMap<_, _>>`

**Comment by nickfarrow** [comment]
Looks good!

This risk was introduced by us being lazy not passing our public poly into finish keygen code. Just receiving a `FinishKeygen` message from the coordinator which contains all the public polys including our own, 

Others will likely do the same, so definitely worth highlighting.

It does make `new_keygen` slightly clunkier, skipping the check with `BTreeMap::new() or Default::default()`, but i think this is probably nicer than an `Option<BTreeMap<_, _>>`

File: schnorr_fun/src/frost.rs (hunk 1 above)
Comment (LLFourn) [reply schnorr_fun/src/frost.rs:687]: No it allows you to contribute multiple secret polynomials. Should probably put on the docs that it doesn't make sense to design a key genration protocol where you do this.

**Comment by LLFourn** [comment]
Thanks @nickfarrow.

> It does make `new_keygen` slightly clunkier, skipping the check with `BTreeMap::new() or Default::default()`, but i think this is probably nicer than an `Option<BTreeMap<_, _>>`

Clunkyness is a feature as it inspires the reader to look at the documentation!

**Pull request merged**